framework/src/media/HttpStream.cpp

```diff
@@ -0,0 +1,188 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************/
+
+#include <curl/curl.h>
+#include <curl/easy.h>
+#include <debug.h>
+
+#include "HttpStream.h"
+
+namespace media {
+namespace stream {
+
+#define SET_OPTION(curl, opt, val) \
+	do {\
+		auto result = curl_easy_setopt(curl, opt, val);\
+		if (result != CURLE_OK) {\
+			meddbg("curl_easy_setopt(%d...) failed!\n", (int)opt);\
+			return false;\
+		}\
+	} while (0)
+
+int HttpStream::mInitialzeCount = 0;
+
+std::shared_ptr<HttpStream> HttpStream::create()
+{
+	std::shared_ptr<HttpStream> stream(new HttpStream());
+	if (!stream->init()) {
+		meddbg("HttpStream instance create failed!\n");
+		return nullptr;
+	}
+
+	return stream;
+}
+
+HttpStream::HttpStream() :
+	mCurl(nullptr),
+	mHttpHeaders(nullptr)
+{
+}
+
+HttpStream::~HttpStream()
+{
+	cleanup();
+}
+
+bool HttpStream::addHeader(const std::string &header) {
+	mHttpHeaders = curl_slist_append(mHttpHeaders, header.c_str());
+	if (!mHttpHeaders) {
+		meddbg("curl_slist_append failed, header %s\n", header.c_str());
+		return false;
+	}
+
+	return true;
+}
+
+bool HttpStream::setHeaderCallback(CallbackFunc callback, void *userdata)
+{
+	SET_OPTION(mCurl, CURLOPT_HEADERFUNCTION, callback);
+	SET_OPTION(mCurl, CURLOPT_HEADERDATA, userdata);
+	return true;
+}
+
+bool HttpStream::setWriteCallback(CallbackFunc callback, void *userdata)
+{
+	SET_OPTION(mCurl, CURLOPT_WRITEFUNCTION, callback);
+	SET_OPTION(mCurl, CURLOPT_WRITEDATA, userdata);
+	return true;
+}
+
+bool HttpStream::setReadCallback(CallbackFunc callback, void *userdata)
+{
+	SET_OPTION(mCurl, CURLOPT_READFUNCTION, callback);
+	SET_OPTION(mCurl, CURLOPT_READDATA, userdata);
+	return true;
+}
+
+bool HttpStream::init()
+{
+	if (mInitialzeCount == 0) {
+		if (curl_global_init(CURL_GLOBAL_ALL) != CURLE_OK) {
+			meddbg("curl_global_init failed\n");
+			return false;
+		}
+	}
+	mInitialzeCount++;
+	mInitialzeFlag = true;
+
+	mCurl = curl_easy_init();
+	if (mCurl == nullptr) {
+		meddbg("HttpStream init failed!\n");
+		return false;
+	}
+
+	return true;
+}
+
+void HttpStream::cleanup()
+{
+	if (mHttpHeaders) {
+		curl_slist_free_all(mHttpHeaders);
+		mHttpHeaders = nullptr;
+	}
+
+	if (mCurl != nullptr) {
+		curl_easy_cleanup(mCurl);
+		mCurl = nullptr;
+	}
+
+	if (mInitialzeFlag) {
+		mInitialzeCount--;
+		if (mInitialzeCount == 0) {
+			curl_global_cleanup();
+		}
+		mInitialzeFlag = false;
+	}
+}
+
+bool HttpStream::perform(CURLoption updown, const char *url)
+{
+	SET_OPTION(mCurl, updown, 1L);
```

Comment:
I think 'updown' is bit ambiguous.
How about set_opt(*GET/*UPLOAD) directly in 'download' and 'upload'

Reply:
yes, I changed it, thx.